### PR TITLE
Swift option added for pfioh

### PR DESCRIPTION
**Pull request by awalkaradi95moc**
@danmcp

### CODE REVIEW

**Comment by awalkaradi95moc** [comment]
@danmcp getData() for Swift has been added.

File: docker-entrypoint.py [hunk 1]
```diff
@@ -38,7 +38,7 @@ def pfioh_do(args, unknown):
 
     str_otherArgs   = ' '.join(unknown)
 
-    str_CMD = "/usr/bin/pfioh --forever %s" % (str_otherArgs)
+    str_CMD = "/usr/bin/pfioh --forever --httpResponse --storeBase=/tmp/share --createDirsAsNeeded"
```
Comment (danmcp) [inline docker-entrypoint.py]: What's this about?

File: pfioh/pfioh.py [hunk 2]
```diff
@@ -173,84 +178,24 @@ def do_GET_withCompression(self, d_msg):
...
-            d_ret['zip']        = d_fio
-            d_ret['status']     = d_fio['status']
-            d_ret['msg']        = d_fio['msg']
-            d_ret['timestamp']  = '%s' % datetime.datetime.now()
-            if not d_ret['status']:
-                self.qprint("An error occurred during the zip operation:\n%s" % d_ret['stdout'],
-                            comms = 'error')
-                self.ret_client(d_ret)
-                return d_ret
+        if b_swiftStorage:
```
Comment (danmcp) [inline pfioh/pfioh.py]: At a min this should be abstracted behind a method.  Would a statically initialized instance make more sense though?

File: docker-entrypoint.py (hunk 1 above)
Comment (awalkaradi95moc) [inline docker-entrypoint.py]: This I had changed long back when we needed pfioh or pman service to start according to changes we made(like --createDirasNeeded was added). But later you added the command to the dc directly and so this is not needed in Dockerfile. I must have forgotten to make the change locally and commit it on my github repo. I will remove this and do a force commit again.

File: pfioh/pfioh.py [hunk 3]
```diff
@@ -86,10 +86,14 @@ def remoteLocation_resolve(self, d_remote):
         """
         b_status        = False
         str_remotePath  = ""
-        if 'path' in d_remote.keys():
+
+        if b_swiftStorage:
```
Comment (danmcp) [inline pfioh/pfioh.py]: Maybe add a TODO to remove this once the path option is removed.

Comment (awalkaradi95moc) [inline pfioh/pfioh.py]: Yes, adding a TODO.

File: pfioh/pfioh.py [hunk 4]
```diff
@@ -656,6 +613,7 @@ def do_POST_withCopy(self, d_meta):
                 shutil.copy2(str_clientPath, str_serverPath)
             d_ret['copytree']       = b_copyTree
             d_ret['copyfile']       = b_copyFile
+
```
Comment (danmcp) [inline pfioh/pfioh.py]: why?

File: pfioh/pfioh.py [hunk 5]
```diff
@@ -796,7 +754,10 @@ def do_POST_withCompression(self, **kwargs):
         #
         # d_meta              = json.loads(d_form['d_meta'])
         fileContent         = d_form['local']
+        self.qprint('**********fileContent*********** %s' %fileContent)
```
Comment (danmcp) [inline pfioh/pfioh.py]: I don't think you want to leave the debug statements.

File: pfioh/store-options-handler.py [hunk 6]
```diff
@@ -0,0 +1,322 @@
+"""
```
Comment (danmcp) [inline pfioh/store-options-handler.py]: Why are swift and mountDir in the same file?

File: pfioh/store-options-handler.py [hunk 7]
```diff
@@ -0,0 +1,322 @@
...
+
+        #Headers 
+        d_fio= {
+	        'msg':              'Unzip operation successful',
+	        'status':           True,
+	        'path':             'Swift Storage',
+	        'zipmode':          'zip',
+	        'timestamp':        '%s' % datetime.datetime.now()
+    	}
+    	d_ret['unzip']  = d_fio
```
Comment (danmcp) [inline pfioh/store-options-handler.py]: code is not lining up

File: pfioh/store-options-handler.py [hunk 8]
```diff
@@ -0,0 +1,322 @@
...
+
+
+	def storeData(self, **kwargs):
+		"""
+		Creates an object of the file and stores it into the container as key-value object 
+		"""
+
+		for k,v in kwargs.items():
+			if k == 'file_content':   fileContent = v
+            if k == 'Path': 	  key= v
```
Comment (danmcp) [inline pfioh/store-options-handler.py]: code is not lining up

File: pfioh/store-options-handler.py [hunk 9]
```diff
@@ -0,0 +1,322 @@
...
+
+		for k,v in kwargs.items():
+			if k== 'path': str_fileToProcess= v
+			if k== 'is_zip': is_zip= v
+			if k== 'encoding': str_encoding= v
+			if k== 'cleanup': b_cleanup= v
+			if k== 'd_ret': d_ret= v
+
+		#Zipping
+		if b_zip:
```
Comment (danmcp) [inline pfioh/store-options-handler.py]: code is not lining up

File: pfioh/store-options-handler.py [hunk 10]
```diff
@@ -0,0 +1,322 @@
...
+	def __init__(self, *args, **kwargs):
+		StoreHandler.__init__(self, *args, **kwargs)
+
+
+	def storeData(self, **kwargs):
+		"""
+		Stores the file/directory at the specified location
+		"""
+
+		for k,v in kwargs.items():
```
Comment (danmcp) [inline pfioh/store-options-handler.py]: code is not lining up

File: pfioh/pfioh.py (hunk 5 above)
Comment (awalkaradi95moc) [inline pfioh/pfioh.py]: Yes, I was going to keep them until I'm finished testing everything.

File: pfioh/store-options-handler.py (hunk 8 above)
Comment (awalkaradi95moc) [inline pfioh/store-options-handler.py]: Fixing it. Strangely, sublime text shows proper indentation, however when I open it in gedit the indentation is messed up.

Comment (danmcp) [inline pfioh/store-options-handler.py]: @awalkaradi95moc You might want to set up sublime to convert tabs to spaces.  I am guessing that's your issue.

Comment (awalkaradi95moc) [inline pfioh/store-options-handler.py]: Okay, done that. Thank you!

**Comment by awalkaradi95moc** [comment]
Please ignore the 2nd commit. I intended to only push on my github repo and not add in the pull request.

File: pfioh/pfioh.py (hunk 4 above)
Comment (awalkaradi95moc) [inline pfioh/pfioh.py]: I didn't make any changes there or in any copy functions.

Comment (danmcp) [inline pfioh/pfioh.py]: It looks like you added a line break

File: pfioh/store-options-handler.py (hunk 6 above)
Comment (danmcp) [inline pfioh/store-options-handler.py]: I would recommend that so when looking at it from the perspective of one implementation you don't get mixed up.

File: pfioh/pfioh.py [hunk 11]
```diff
@@ -151,7 +156,7 @@ def do_GET_withCompression(self, d_msg):
         d_compress          = d_transport['compress']
         d_ret               = {}
 
-        str_serverPath      = self.remoteLocation_resolve(d_remote)['path']
+        str_serverPath      = self.remoteLocation_resolve(d_remote, )['path']
```
Comment (danmcp) [inline pfioh/pfioh.py]: What is this about?

File: pfioh/pfioh.py [hunk 12]
```diff
@@ -163,7 +168,8 @@ def do_GET_withCompression(self, d_msg):
         # b_zip               = True
 
         str_encoding        = 'base64'
-
+        if 'encoding' in d_compress: str_encoding    = d_compress['encoding']
```
Comment (danmcp) [inline pfioh/pfioh.py]: Remove the extra spacing before the =

File: pfioh/pfioh.py [hunk 13]
```diff
@@ -173,84 +179,18 @@ def do_GET_withCompression(self, d_msg):
...
-
-        if b_cleanup:
-            if b_zip:
-                self.qprint("Removing '%s'..." % (str_zipFile), comms = 'status')
-                if os.path.isfile(str_zipFile):     os.remove(str_zipFile)
-            if str_encoding == 'base64':
-                self.qprint("Removing '%s'..." % (str_base64File), comms = 'status')
-                if os.path.isfile(str_base64File):  os.remove(str_base64File)
+        
+        d_ret = self.getData(path=str_fileToProcess, is_zip= b_zip,
```
Comment (danmcp) [inline pfioh/pfioh.py]: These two lines don't line up well.

File: pfioh/pfioh.py [hunk 14]
```diff
@@ -336,9 +276,13 @@ def do_GET(self):
             d_ret = self.do_GET_withCompression(d_msg)
             return d_ret
 
-        if 'copy'           in d_transport:
-            d_ret = self.do_GET_withCopy(d_msg)
-            return d_ret
+        if 'copy'           in d_transport :
+            if not b_swiftStorage:
```
Comment (danmcp) [inline pfioh/pfioh.py]: reverse the order here so you don't have to do 'if not'

File: pfioh/pfioh.py [hunk 15]
```diff
@@ -576,7 +520,7 @@ def do_POST(self, **kwargs):
         if 'action' in d_msg:
             self.qprint("verb: %s detected." % d_msg['action'], comms = 'status')
             if 'Path' not in d_msg['action']:
-                str_method      = '%s_process' % d_msg['action']
+                str_method      = '%s_process'  % d_msg['action']
```
Comment (danmcp) [inline pfioh/pfioh.py]: Why?

File: pfioh/pfioh.py [hunk 16]
```diff
@@ -599,8 +543,16 @@ def do_POST(self, **kwargs):
                     form    = form,
                     d_form  = d_form
                 )
-            if 'copy' in d_transport:
-                d_ret   = self.do_POST_withCopy(d_meta)
+            if 'copy' in d_transport :
+                if not b_swiftStorage:
```
Comment (danmcp) [inline pfioh/pfioh.py]: Again with the if not

File: pfioh/pfioh.py [hunk 17]
```diff
@@ -599,8 +543,16 @@ def do_POST(self, **kwargs):
...
+                    d_ret   = self.do_POST_withCopy(d_meta)
```
Comment (danmcp) [inline pfioh/pfioh.py]: Remove the extra spacing before the =

File: pfioh/pfioh.py [hunk 18]
```diff
@@ -599,8 +543,16 @@ def do_POST(self, **kwargs):
...
+                else:
+                    d_ret = self.do_POST_withCompression(
```
Comment (danmcp) [inline pfioh/pfioh.py]: might as well make the = after d_ret line up with the ones below.

File: pfioh/store-options-handler.py [hunk 19]
```diff
@@ -0,0 +1,322 @@
...
+        for k,v in kwargs.items():
+            if k== 'path': key= v
+            if k== 'is_zip': b_zip= v
+            if k== 'encoding': str_encoding= v
+            if k== 'cleanup': b_cleanup= v
+            if k== 'd_ret': d_ret= v
+
+
+        try:
+            swiftConnection= self._initiateSwiftConnection()
```
Comment (danmcp) [inline pfioh/store-options-handler.py]: should have spaces before and after = and == throughout this file.

File: pfioh/store_options.py [hunk 20]
```diff
@@ -0,0 +1,348 @@
...
+
+    def _getSwiftCredentials(self, configPath):
+        """
+        Initializes the swift storage parameters by accessing configmap
+        """
+
+        with open(configPath,'r') as f:
+            line= f.read()
+
+        credentialsDict= ast.literal_eval(line)
```
Comment (danmcp) [inline pfioh/store_options.py]: Why not use a normal config format?

File: pfioh/store_options.py [hunk 21]
```diff
@@ -0,0 +1,348 @@
...
+
+        try:
+            swiftConnection.put_object(containerName, key , contents=value, content_type='text/plain')
+            self.qprint('Object added with key %s' %key)
+
+        except Exception as exp:
+            self.qprint(exp)
+
+
+    def _getObject(self, swiftConnection, containerName, key, b_Delete):
```
Comment (danmcp) [inline pfioh/store_options.py]: The name b_Delete is inconsistent.

File: pfioh/store_options.py [hunk 22]
```diff
@@ -0,0 +1,348 @@
...
+        return swiftDataObject
+
+
+    def storeData(self, **kwargs):
+        """
+        Creates an object of the file and stores it into the container as key-value object 
+        """
+
+        for k,v in kwargs.items():
+            if k == 'file_content':   fileContent = v
```
Comment (danmcp) [inline pfioh/store_options.py]: The assignments on the right should line up.

File: pfioh/store_options.py [hunk 23]
```diff
@@ -0,0 +1,348 @@
...
+        objectInformation= dataObject[0]
+        objectValue= dataObject[1]
+        fileContent= objectValue
+
+        #Zipping
+        if b_zip:
+            try:
+                unZippedFile = StringIO(objectValue)
+                fileContent = zipfile.Zipfile(unzippedFile,"w")
+                self.qprint('Zip sucssful')
```
Comment (danmcp) [inline pfioh/store_options.py]: successful

File: pfioh/store_options.py~ [hunk 24]
```diff
@@ -0,0 +1,330 @@
...
+        """
+
+        passwordAuth = v3.Password(auth_url= osAuthUrl,
+                            user_domain_name= 'default',
+                            username= username, password= password,
+                            project_domain_name= osProjectDomain,
+                            project_name= osProjectName,
+                            unscoped= False)
+
+        scopedSession = session.Session(auth= passwordAuth)
```
Comment (danmcp) [inline pfioh/store_options.py~]: remove space after auth=

File: pfioh/store_options.py~ [hunk 25]
```diff
@@ -0,0 +1,330 @@
...
+        d_ret['status']     = d_fio['status']
+        d_ret['msg']        = d_fio['msg']
+        d_ret['timestamp']  = '%s' % datetime.datetime.now()
+
+            
+        #Encoding
+        if str_encoding=='base64':
+            try:
+                fileContent            = base64.b64encode(fileContent)
+                self.qprint('Base64 encoding sucssful')
```
Comment (danmcp) [inline pfioh/store_options.py~]: successful

File: pfioh/store_options.py~ [hunk 26]
```diff
@@ -0,0 +1,330 @@
...
+        d_ret['zip']        = d_fio
+        d_ret['status']     = d_fio['status']
+        d_ret['msg']        = d_fio['msg']
+        d_ret['timestamp']  = '%s' % datetime.datetime.now()
+
+            
+        #Encoding
+        if str_encoding=='base64':
+            try:
+                fileContent            = base64.b64encode(fileContent)
```
Comment (danmcp) [inline pfioh/store_options.py~]: remove extra spaces

File: pfioh/store_options.py (hunk 20 above)
Comment (awalkaradi95moc) [inline pfioh/store_options.py]: I am not exactly sure about the format. I created the file like a simple dictionary. Should it be like .kube/config?

**Comment by awalkaradi95moc** [comment]
Hi Dan, the last commit is the merged commit of all previous commits after rebasing. Not sure why it didn't delete the previous ones, I'm looking into it.

File: bin/pfioh [hunk 27]
```diff
@@ -21,11 +21,42 @@ import  pudb
...
 try:
-    import  pfioh
+    import pfioh
+    import mount_dir
+    import swift_store
 except:
     sys.path.insert(1, os.path.join(get_python_lib(True, True), 'site-packages'))
     sys.path.insert(1, os.path.join(get_python_lib(True, True), 'site-packages/pfioh'))
-    import  pfioh
+    sys.path.insert(1, os.path.join(get_python_lib(True, True), 'site-packages/mount_dir'))
```
Comment (danmcp) [inline bin/pfioh]: This is only relevant if we are building mount_dir and swift_store as separate packages, which I don't think you are.

File: pfioh/pfioh.py [hunk 28]
```diff
@@ -525,16 +468,50 @@ def hello_process(self, *args, **kwargs):
                 b_status                        = True
 
         return { 'stdout': { 
-                            'd_ret':   d_ret,
+                            'd_ret':   d_ret,
```
Comment (danmcp) [inline pfioh/pfioh.py]: why?

File: pfioh/pfioh.py [hunk 29]
```diff
@@ -525,16 +468,50 @@ def hello_process(self, *args, **kwargs):
...
+        """
+        Return headers of the request
+        """
+        
+        self.qprint('headers= %s' %self.headers)
+        return self.headers['content-length']     
+
+    def rfileRead(self, length):
+        """
+        Return the contets of the file transmitted
```
Comment (danmcp) [inline pfioh/pfioh.py]: contents

File: pfioh/pfioh.py [hunk 30]
```diff
@@ -525,16 +468,50 @@ def hello_process(self, *args, **kwargs):
...
+    def unpackForm(self, form, d_form):
+        """
+        Load the json request
+        """
+
+        self.qprint("Unpacking multi-part form message...", comms = 'status')
+        for key in form:
+            self.qprint("\tUnpacking field '%s..." % key, comms = 'status')
+            d_form[key]     = form.getvalue(key)
+        d_msg               = json.loads((d_form['d_msg']))
```
Comment (danmcp) [inline pfioh/pfioh.py]: making this = line up looks strange

File: bin/pfioh (hunk 27 above)
Comment (awalkaradi95moc) [inline bin/pfioh]: Yes, I have separated them into two different files. Also in #200 and #202, I am passing the two classes from these two packages. This I have added so as to improve the previous implementation where I was instantiating the classes in pfioh.py with if else statements.

File: pfioh/pfioh.py [hunk 31]
```diff
@@ -818,63 +804,56 @@ def do_POST_withCompression(self, **kwargs):
...
                 d_ret['decode']['msg']      = 'base64 decode successful!'
-            except:
+            except Exception as err:
                 d_ret['decode']['status']   = False
                 d_ret['decode']['msg']      = 'base64 decode unsuccessful!'
                 self.ret_client(d_ret)
                 self.qprint(d_ret, comms = 'tx')
+                self.qprint(err)
                 return d_ret
-            d_ret['decode']['timestamp']  = '%s' % datetime.datetime.now()
```
Comment (danmcp) [inline pfioh/pfioh.py]: Why is this removed?

File: pfioh/pfioh.py [hunk 32]
```diff
@@ -818,63 +804,56 @@ def do_POST_withCompression(self, **kwargs):
...
         self.end_headers()
 
         d_ret['User-agent'] = self.headers['user-agent']
 
         # self.ret_client(d_ret)
         self.qprint(d_ret, comms = 'tx')
-
         return d_ret
 
+    def storeData(self, **kwargs):
```
Comment (danmcp) [inline pfioh/pfioh.py]: Why are you doing this?

File: pfioh/pfioh.py (hunk 31 above)
Comment (awalkaradi95moc) [inline pfioh/pfioh.py]: Must have got deleted while splitting into mountdir and swiftstore. Added it back.

File: pfioh/pfioh.py (hunk 32 above)
Comment (awalkaradi95moc) [inline pfioh/pfioh.py]: I thought 2 lines space between functions looks more readable. Should I keep a single line space only?

File: pfioh/swift_store.py [hunk 33]
```diff
@@ -0,0 +1,218 @@
...
+    def _initiateSwiftConnection(self, **kwargs):
+        """
+        Initiates a Swift connection and returns a Swift Connection Object
+        """
+
+        for k,v in kwargs:
+            if k == 'configPath': configPath= v
+
+        #Delete the path
+        configPath= '/home/aditya/Chris/temp-scripts/swift/swift-credentials.cfg'
```
Comment (danmcp) [inline pfioh/swift_store.py]: Need to document how to use this.  Particularly as a secret with OpenShift.

File: pfioh/pfioh.py (hunk 32 above)
Comment (awalkaradi95moc) [inline pfioh/pfioh.py]: I am basically declaring that function as abstract. The function is overridden in both MountDir and SwiftStore. 
I referred this : https://stackoverflow.com/a/4383103/5871403

File: bin/pfioh (hunk 27 above)
Comment (danmcp) [inline bin/pfioh]: When you install pfioh, does this dir:

site-packages/mount_dir

exist?

File: pfioh/pfioh.py (hunk 32 above)
Comment (danmcp) [inline pfioh/pfioh.py]: To be clear, does something not work if this is removed?

Comment (awalkaradi95moc) [inline pfioh/pfioh.py]: No, this is just in case that some base class has not implemented the method, we should come to know that the method has to be implemented. Java throws an error if the base class has not implemented the method, this is kind of doing the same.  
When it does Dynamic method dispatch, it will first search for the method in the class whose object I have used to call the method with. If it does not find it there, it will jump to parent class to look for that method.

Comment (danmcp) [inline pfioh/pfioh.py]: Nit:

> it will first search for the method in the class whose object I have used to call the method with

Technically I think you mean it will make the call from the object that was instantiated.  Which might not be the class that is making the call.

I was really just asking if it was required in some way.  Didn't know why you would have done it in the first place unless it was a pattern you were familiar with.  I can't say I love the pattern as it would already give an error if it wasn't implemented.  And it just adds cruft to the code.  But not being a long time python coder, I can't comment on whether it's idiomatic.
